EE-7979 - Document SDK logging capabilities (#163)
mulesoft/docs-mule-sdk: modules/ROOT/pages/troubleshooting.adoc

> = Troubleshoot SDK Components - Mule 4

Reviewer: ```suggestion
= Troubleshoot SDK Components and Operations - Mule 4
```
See below

ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:keywords: mule, sdk, troubleshooting, logging

> Troubleshoot SDK components to monitor and get more information about your application.

Reviewer: = Troubleshoot SDK Components - Mule 4
<metadata>
Use logging to troubleshoot SDK components; use _verbose logging_ to test and troubleshoot operations. 
...
== Logging (or Troubleshooting SDK Components?)
Use logging to troubleshoot the following SDK components:
...
== Verbose Logging (or Testing and Troubleshooting Operations?)
Because of the high impact to application performance, limit verbose logging to testing and troubleshooting operations. Enable verbose logging in Anypoint Studio or Anypoint Runtime Manager if using CloudHub APIs.
In both cases, the application's log file logs the information.

=== Enable Verbose Logging in Anypoint Studio
...
=== Enable Verbose Logging in Anypoint Runtime Manager
...

Reviewer: Rest is copyedited as-is

== Logging

Use logging to troubleshoot SDK components. SDK supports logging for the following components:

[%header,cols="2,3"]
|===
| Component Name | Logging Capabilities
| <<sources#, Source>> | Logs information about the source lifecycle, such as when it starts and ends.

> In case of error and reconnection, it provides the full stack trace associated with the raised exception.

Reviewer: ```suggestion
In case of error and reconnection, the log provides the full stack trace associated with the raised exception.
```

> For cluster environments, it provides source status information, which indicates whether it is the primary source or not.

Reviewer: ```suggestion
For cluster environments, the log provides source status information, which indicates whether the source is primary or not.
```

Reviewer: GLOBALLY in writing, review your use of "it" and consider replacing with the actual noun being documented.

> | <<polling-sources#, Polling Source>> | Logs information about accepted and rejected items by the source.

Reviewer: ```suggestion
| <<polling-sources#, Polling Source>> | Logs information about items accepted and rejected by the source.
```

> In case of rejection, it logs information about its cause.

Reviewer: ```suggestion
In the case of rejection, the log contains information about the cause of the rejection.
```

| <<polling-sources#_watermarking, Polling Source Watermarking>> | Logs information about watermarking handling on polling sources.

> Because this uses low level logic, enable trace logging only when debugging or troubleshooting is necessary.

Reviewer: ```suggestion
Because this process uses low-level logic, enable trace logging only when debugging or troubleshooting is necessary.
```
....but, isn't that the case with all troubleshooting, that it is necessary? If not, we need to clarify whether this topic is really about trouble or just about obtaining info.

| <<connections#_pooling-strategy, Pooling Connection Provider>> | Logs information about connection management for pooling connection providers.

> It provides connection pool status information as well as connection handling, which indicates when the pool retrieves and returns connections.

Reviewer: ```suggestion
This log provides connection pool status information as well as connection handling, which indicates when the pool retrieves and returns connections.
```

|===

== Verbose Logging

Use verbose logging to troubleshoot SDK components to get information regarding their status or behavior.
Because of the high impact to application performance, limit verbose logging to testing and troubleshooting operations.

> Enable verbose logging in Anypoint Studio or Anypoint Runtime Manager if using CloudHub APIs.

Reviewer: Does "if using CloudHub APIs" modify both apps or just ARM? Clarify.

In both cases, the application's log file logs the information.

=== Enable Verbose Logging in Anypoint Studio

> The following steps describe how to enable verbose logging using Anypoint Studio:

Reviewer: Is there anything to add about why the audience wants to use Studio vs. ARM?

. Access Anypoint Studio and navigate to the *Package Explorer* view.
. Open your application's project name.
. Open the `src/main/resources` path folder.
. Open the `log4j2.xml` file inside the folder.

+
image:enable-sdk-logging-studio.png[enable-sdk-logging-studio]
+

. Add an `<AsyncLogger>` tag inside the `<Loggers>` tag:

+
[source,xml,linenums]
----
<Loggers>
    ...
    <AsyncLogger name="<component's fully qualified name>" level="<logging level>"/>
    ...
</Loggers>
----
+

> Use the following table as reference to specify the component's fully qualified name:

Reviewer: ```suggestion
Use the following table as a reference for specifying the component's fully qualified name:
```

[%header,cols="2,2,1"]
|===
| Name | Fully Qualified Name | Log Level
| Source|`org.mule.runtime.module.extension.internal.runtime.source.ExtensionMessageSource`| `DEBUG`
| Polling Source|`org.mule.runtime.module.extension.internal.runtime.source.poll.PollingSourceWrapper`| `DEBUG`
| Polling Source Watermarking|`org.mule.runtime.module.extension.internal.runtime.source.poll.PollingSourceWrapper`| `TRACE`
| Pooling Connection Provider|`org.mule.runtime.core.internal.connection.PoolingConnectionHandler`
`org.mule.runtime.core.internal.connection.PoolingConnectionManagementStrategy`| `DEBUG`
|===

The following example shows how to add a fully qualified name to the `log4j2.xml` file to enable verbose logging for Source:

[source,xml,linenums]
----
<Loggers>
   ...
    <!-- Mule logger -->
    <AsyncLogger name="org.mule.runtime.core.internal.processor.LoggerMessageProcessor" level="INFO"/>
    <AsyncLogger name="org.mule.runtime.module.extension.internal.runtime.source.ExtensionMessageSource" level="DEBUG"/>
    <AsyncRoot level="INFO">
        <AppenderRef ref="file" />
    </AsyncRoot>
</Loggers>
----

=== Enable Verbose Logging in Anypoint Runtime Manager

> The following steps describe how to enable verbose logging using Anypoint Runtime Manager:

Reviewer: ditto why ARM vs. Studio?

. Access Anypoint Runtime Manager.
. Select the desired application.
. Go to *Settings* on the left pane.
. Go to the *Logging* tab.

+
image:enable-sdk-logging-runtime-manager.png[sdk-verbose-logging-runtime-manager]
+

. Add the component's fully qualified name and the logging level (DEBUG or TRACE).

> Refer to the previous section on how to specify the component's fully qualified name.

Reviewer: We don't want to provide a locator like "previous section" because it creates a locational dependence and blocks reuse. Instead, maybe something like "Refer to the table that...."
